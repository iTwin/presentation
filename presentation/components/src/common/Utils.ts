/*---------------------------------------------------------------------------------------------
* Copyright (c) 2019 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
/** @module Core */

import * as React from "react";
import { I18NNamespace } from "@bentley/imodeljs-i18n";
import { Presentation } from "@bentley/presentation-frontend";

/**
 * An interface of something that has a priority.
 * @internal
 */
export interface IPrioritized {
  priority: number;
}

/**
 * A sorting algorithm for `Array.sort` that sorts items by priority.
 * Higher priority items appear first in the list.
 *
 * @internal
 */
export const prioritySortFunction = (a: IPrioritized, b: IPrioritized): number => {
  if (a.priority > b.priority)
    return -1;
  if (a.priority < b.priority)
    return 1;
  return 0;
};

let localizationNamespace: I18NNamespace | undefined;
/**
 * Translate a string with the specified id from `PresentationComponents`
 * localization namespace. The `stringId` should not contain namespace - it's
 * prepended automatically.
 *
 * @internal
 */
export const translate = async (stringId: string): Promise<string> => {
  const localizationNamespaceName = "PresentationComponents";
  if (!localizationNamespace) {
    localizationNamespace = Presentation.i18n.registerNamespace(localizationNamespaceName);
  }
  await localizationNamespace.readFinished;
  stringId = `${localizationNamespaceName}:${stringId}`;
  return Presentation.i18n.translate(stringId);
};

/**
 * Creates a display name for the supplied component
 * @internal
 */
export const getDisplayName = <P>(component: React.ComponentType<P>): string => {
  if (component.displayName)
    return component.displayName;
  if (component.name)
    return component.name;
  return "Component";
};
