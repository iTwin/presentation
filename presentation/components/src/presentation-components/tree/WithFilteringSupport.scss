/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/

.filteredTree {

  position: relative;
  width: 100%;
  height: 100%;

  .filteredTreeOverlay {
    position: absolute;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    z-index: 9999;
    background-color: rgba(255,255,255,.8);
  }

}
