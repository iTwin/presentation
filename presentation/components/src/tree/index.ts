/*---------------------------------------------------------------------------------------------
* Copyright (c) 2018 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
/** @module Tree */

export { default as PresentationTreeDataProvider } from "./DataProvider";
export { default as IPresentationTreeDataProvider } from "./IPresentationTreeDataProvider";
export { default as withUnifiedSelection, SelectionTarget } from "./WithUnifiedSelection";
export { default as withFilteringSupport } from "./WithFilteringSupport";
