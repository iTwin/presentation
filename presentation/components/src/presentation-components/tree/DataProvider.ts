/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
/** @packageDocumentation
 * @module Tree
 */

import memoize from "micro-memoize";
import { IDisposable, Logger } from "@bentley/bentleyjs-core";
import { IModelConnection } from "@bentley/imodeljs-frontend";
import { HierarchyRequestOptions, NodeKey, NodePathElement, Ruleset } from "@bentley/presentation-common";
import { Presentation } from "@bentley/presentation-frontend";
import { DelayLoadedTreeNodeItem, PageOptions, TreeNodeItem } from "@bentley/ui-components";
import { RulesetRegistrationHelper } from "../common/RulesetRegistrationHelper";
import { IPresentationTreeDataProvider } from "./IPresentationTreeDataProvider";
import { createTreeNodeItems, pageOptionsUiToPresentation, PRESENTATION_TREE_NODE_KEY } from "./Utils";

/**
 * Properties for creating a `PresentationTreeDataProvider` instance.
 * @public
 */
export interface PresentationTreeDataProviderProps {
  /** IModel to pull data from. */
  imodel: IModelConnection;
  /** Id of the ruleset to use when requesting content or a ruleset itself. */
  ruleset: string | Ruleset;
  /**
   * Paging size for obtaining nodes.
   *
   * Presentation data providers, when used with paging, have ability to save one backend request for size / count. That
   * can only be achieved when `pagingSize` property is set on the data provider and it's value matches size which is used when
   * requesting nodes. To help developers notice this problem, data provider emits a warning similar to this:
   * ```
   * PresentationTreeDataProvider.pagingSize doesn't match pageOptions in PresentationTreeDataProvider.getNodes call. Make sure you set PresentationTreeDataProvider.pagingSize to avoid excessive backend requests.
   * ```
   * To fix the issue, developers should make sure the page size used for requesting data is also set for the data provider:
   * ```TS
   * const pagingSize = 10;
   * const provider = new TreeDataProvider({imodel, ruleset, pagingSize});
   * // only one backend request is made for the two following requests:
   * provider.getNodesCount();
   * provider.getNodes({ start: 0, size: pagingSize });
   * ```
   */
  pagingSize?: number;
}

/**
 * Presentation Rules-driven tree data provider.
 * @public
 */
export class PresentationTreeDataProvider implements IPresentationTreeDataProvider, IDisposable {
  private _imodel: IModelConnection;
  private _rulesetRegistration: RulesetRegistrationHelper;
  private _pagingSize?: number;

  /** Constructor. */
  public constructor(props: PresentationTreeDataProviderProps) {
    this._rulesetRegistration = new RulesetRegistrationHelper(props.ruleset);
    this._imodel = props.imodel;
    this._pagingSize = props.pagingSize;
  }

  /** Destructor. Must be called to clean up.  */
  public dispose() {
    this._rulesetRegistration.dispose();
  }

  /** Id of the ruleset used by this data provider */
  public get rulesetId(): string { return this._rulesetRegistration.rulesetId; }

  /** [[IModelConnection]] used by this data provider */
  public get imodel(): IModelConnection { return this._imodel; }

  /**
   * Paging options for obtaining nodes.
   * @see `PresentationTreeDataProviderProps.pagingSize`
   */
  public get pagingSize(): number | undefined { return this._pagingSize; }
  public set pagingSize(value: number | undefined) { this._pagingSize = value; }

  /** Called to get extended options for node requests */
  private createRequestOptions(): HierarchyRequestOptions<IModelConnection> {
    return {
      imodel: this._imodel,
      rulesetOrId: this._rulesetRegistration.rulesetId,
    };
  }

  /**
   * Returns a [[NodeKey]] from given [[TreeNodeItem]].
   * **Warning:** the `node` must be created by this data provider.
   */
  public getNodeKey(node: TreeNodeItem): NodeKey {
    return (node as any)[PRESENTATION_TREE_NODE_KEY];
  }

  /**
   * Returns nodes
   * @param parentNode The parent node to return children for.
   * @param pageOptions Information about the requested page of data.
   */
  public async getNodes(parentNode?: TreeNodeItem, pageOptions?: PageOptions): Promise<DelayLoadedTreeNodeItem[]> {
    if (undefined !== pageOptions && pageOptions.size !== this.pagingSize) {
      const msg = `PresentationTreeDataProvider.pagingSize doesn't match pageOptions in PresentationTreeDataProvider.getNodes call.
        Make sure you set PresentationTreeDataProvider.pagingSize to avoid excessive backend requests.`;
      Logger.logWarning("Presentation.Components", msg);
    }

    return (await this._getNodesAndCount(parentNode, pageOptions)).nodes;
  }

  /**
   * Returns the total number of nodes
   * @param parentNode The parent node to return children count for.
   */
  public async getNodesCount(parentNode?: TreeNodeItem): Promise<number> {
    const pageOptions = undefined !== this.pagingSize ? { start: 0, size: this.pagingSize } : undefined;
    return (await this._getNodesAndCount(parentNode, pageOptions)).count;
  }

  private _getNodesAndCount = memoize(async (parentNode?: TreeNodeItem, pageOptions?: PageOptions): Promise<{ nodes: TreeNodeItem[], count: number }> => {
    const requestCount = undefined !== pageOptions && 0 === pageOptions.start && undefined !== pageOptions.size;
    const parentKey = parentNode ? this.getNodeKey(parentNode) : undefined;

    if (!requestCount) {
      const allNodes = await Presentation.presentation.getNodes({ ...this.createRequestOptions(), paging: pageOptionsUiToPresentation(pageOptions) }, parentKey);
      return { nodes: parentNode ? createTreeNodeItems(allNodes, parentNode.id) : createTreeNodeItems(allNodes), count: allNodes.length };
    }

    const nodesResponse = await Presentation.presentation.getNodesAndCount({ ...this.createRequestOptions(), paging: pageOptionsUiToPresentation(pageOptions) }, parentKey);
    return { nodes: parentNode ? createTreeNodeItems(nodesResponse.nodes, parentNode.id) : createTreeNodeItems(nodesResponse.nodes), count: nodesResponse.count };
  }, { isMatchingKey: MemoizationHelpers.areNodesRequestsEqual as any });

  /**
   * Returns filtered node paths.
   * @param filter Filter.
   */
  public getFilteredNodePaths = async (filter: string): Promise<NodePathElement[]> => {
    return Presentation.presentation.getFilteredNodePaths(this.createRequestOptions(), filter);
  }

  /**
   * Loads the hierarchy so on-demand requests and filtering works quicker
   * @alpha Hierarchy loading performance needs to be improved before this becomes publicly available.
   */
  public async loadHierarchy() {
    return Presentation.presentation.loadHierarchy(this.createRequestOptions());
  }

}

class MemoizationHelpers {
  public static areNodesRequestsEqual(lhsArgs: [TreeNodeItem?, PageOptions?], rhsArgs: [TreeNodeItem?, PageOptions?]): boolean {
    if (lhsArgs[0]?.id !== rhsArgs[0]?.id)
      return false;
    if ((lhsArgs[1]?.start ?? 0) !== (rhsArgs[1]?.start ?? 0))
      return false;
    if ((lhsArgs[1]?.size ?? 0) !== (rhsArgs[1]?.size ?? 0))
      return false;
    return true;
  }
}
