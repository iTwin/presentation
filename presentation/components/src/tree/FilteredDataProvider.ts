/*---------------------------------------------------------------------------------------------
* Copyright (c) 2019 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
/** @module Tree */

import { IModelConnection } from "@bentley/imodeljs-frontend";
import { NodePathElement, NodeKey } from "@bentley/presentation-common";
import {
  SimpleTreeDataProvider, SimpleTreeDataProviderHierarchy,
  DelayLoadedTreeNodeItem, TreeNodeItem,
  PageOptions, ActiveMatchInfo,
} from "@bentley/ui-components";
import { createTreeNodeItem } from "./Utils";
import { IPresentationTreeDataProvider } from "./IPresentationTreeDataProvider";
import { memoize } from "lodash";

/**
 * Rules-driven presentation tree data provider that returns filtered results.
 * @internal
 */
export class FilteredPresentationTreeDataProvider implements IPresentationTreeDataProvider {
  private _parentDataProvider: IPresentationTreeDataProvider;
  private _filteredDataProvider: SimpleTreeDataProvider;
  private _filter: string;
  private _filteredResultMatches: Array<{ id: string, matchesCount: number }> = [];

  public constructor(parentDataProvider: IPresentationTreeDataProvider, filter: string, paths: ReadonlyArray<Readonly<NodePathElement>>) {
    this._parentDataProvider = parentDataProvider;
    this._filter = filter;
    const hierarchy: SimpleTreeDataProviderHierarchy = new Map<string | undefined, TreeNodeItem[]>();
    this.createHierarchy(paths, hierarchy);
    this._filteredDataProvider = new SimpleTreeDataProvider(hierarchy);
  }

  public get rulesetId(): string { return this._parentDataProvider.rulesetId; }

  public get imodel(): IModelConnection { return this._parentDataProvider.imodel; }

  public get filter(): string { return this._filter; }

  public get parentDataProvider(): IPresentationTreeDataProvider { return this._parentDataProvider; }

  private createHierarchy(paths: ReadonlyArray<Readonly<NodePathElement>>, hierarchy: SimpleTreeDataProviderHierarchy, parentId?: string) {
    const treeNodes: DelayLoadedTreeNodeItem[] = [];
    for (let i = 0; i < paths.length; i++) {
      const node = createTreeNodeItem(paths[i].node, parentId);

      if (paths[i].filteringData && paths[i].filteringData!.matchesCount)
        this._filteredResultMatches.push({ id: node.id, matchesCount: paths[i].filteringData!.matchesCount });

      if (paths[i].children.length !== 0) {
        this.createHierarchy(paths[i].children, hierarchy, node.id);
        node.hasChildren = true;
        node.autoExpand = true;
      } else {
        delete node.hasChildren;
        delete node.autoExpand;
      }

      treeNodes[i] = node;
    }
    hierarchy.set(parentId, treeNodes);
  }

  public getActiveMatch: (index: number) => ActiveMatchInfo | undefined = memoize((index: number): ActiveMatchInfo | undefined => {
    let activeMatch: ActiveMatchInfo | undefined;
    if (index <= 0)
      return undefined;

    let i = 1;
    for (const node of this._filteredResultMatches) {
      if (index < i + node.matchesCount) {
        activeMatch = {
          nodeId: node.id,
          matchIndex: index - i,
        };
        break;
      }

      i += node.matchesCount;
    }
    return activeMatch;
  });

  /** Count filtering results. Including multiple possible matches within node labels */
  public countFilteringResults(nodePaths: ReadonlyArray<Readonly<NodePathElement>>): number {
    let resultCount = 0;

    // Loops through root level only
    for (const path of nodePaths) {
      if (path.filteringData)
        resultCount += path.filteringData.matchesCount + path.filteringData.childMatchesCount;
    }

    return resultCount;
  }

  public async getNodes(parent?: TreeNodeItem, pageOptions?: PageOptions): Promise<DelayLoadedTreeNodeItem[]> {
    return this._filteredDataProvider.getNodes(parent, pageOptions);
  }

  public async getNodesCount(parent?: TreeNodeItem): Promise<number> {
    return this._filteredDataProvider.getNodesCount(parent);
  }

  public async getFilteredNodePaths(filter: string): Promise<NodePathElement[]> {
    return this._parentDataProvider.getFilteredNodePaths(filter);
  }

  public getNodeKey(node: TreeNodeItem): NodeKey {
    return this._parentDataProvider.getNodeKey(node);
  }

  /** @alpha */
  // istanbul ignore next
  public async loadHierarchy() {
    // the hierarchy is already loaded when this provider is created
  }

}
