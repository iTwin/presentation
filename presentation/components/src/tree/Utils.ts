/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
/** @packageDocumentation
 * @module Tree
 */

import { StyleHelper } from "../common/StyleHelper";
import { CheckBoxState } from "@bentley/ui-core";
import { Node, PageOptions as PresentationPageOptions, LabelDefinition } from "@bentley/presentation-common";
import { DelayLoadedTreeNodeItem, PageOptions as UiPageOptions, ItemStyle, ItemColorOverrides } from "@bentley/ui-components";
import { createLabelRecord } from "../common/Utils";

/** @internal */
export const PRESENTATION_TREE_NODE_KEY = "__presentation-components/key";

/** @internal */
export const createTreeNodeItems = (nodes: ReadonlyArray<Readonly<Node>>, parentId?: string): DelayLoadedTreeNodeItem[] => {
  const list = new Array<DelayLoadedTreeNodeItem>();
  for (const node of nodes)
    list.push(createTreeNodeItem(node, parentId));
  return list;
};

/** @internal */
export const createTreeNodeItem = (node: Readonly<Node>, parentId?: string): DelayLoadedTreeNodeItem => {
  const item: DelayLoadedTreeNodeItem = {
    id: [...node.key.pathFromRoot].reverse().join("/"),
    label: node.label,
    labelDefinition: createLabelRecord(node.labelDefinition ? node.labelDefinition : LabelDefinition.fromLabelString(node.label), "node_label"),
  };
  (item as any)[PRESENTATION_TREE_NODE_KEY] = node.key;

  const style: ItemStyle = {};
  const colorOverrides: ItemColorOverrides = {};
  if (parentId)
    item.parentId = parentId;
  if (node.description)
    item.description = node.description;
  if (node.hasChildren)
    item.hasChildren = true;
  if (node.isExpanded)
    item.autoExpand = true;
  if (node.imageId)
    item.icon = node.imageId;
  if (StyleHelper.isBold(node))
    style.isBold = true;
  if (StyleHelper.isItalic(node))
    style.isItalic = true;
  const foreColor = StyleHelper.getForeColor(node);
  if (foreColor)
    colorOverrides.color = foreColor;
  const backColor = StyleHelper.getBackColor(node);
  if (backColor)
    colorOverrides.backgroundColor = backColor;
  if (node.isCheckboxVisible) {
    item.isCheckboxVisible = true;
    if (node.isChecked)
      item.checkBoxState = CheckBoxState.On;
    if (!node.isCheckboxEnabled)
      item.isCheckboxDisabled = true;
  }
  if (Object.keys(colorOverrides).length > 0)
    style.colorOverrides = colorOverrides;
  if (Object.keys(style).length > 0)
    item.style = style;
  if (node.extendedData)
    item.extendedData = node.extendedData;

  return item;
};

/** @internal */
export const pageOptionsUiToPresentation = (pageOptions?: UiPageOptions): PresentationPageOptions | undefined => {
  if (pageOptions)
    return { ...pageOptions };
  return undefined;
};
