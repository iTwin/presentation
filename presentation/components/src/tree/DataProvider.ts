/*---------------------------------------------------------------------------------------------
* Copyright (c) 2018 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
/** @module Tree */

import * as _ from "lodash";
import { IModelConnection } from "@bentley/imodeljs-frontend";
import { NodeKey, NodePathElement, HierarchyRequestOptions } from "@bentley/presentation-common";
import { Presentation } from "@bentley/presentation-frontend";
import { DelayLoadedTreeNodeItem, TreeNodeItem } from "@bentley/ui-components/lib/tree/TreeDataProvider";
import { PageOptions } from "@bentley/ui-components/lib/common/PageOptions";
import { createTreeNodeItems, pageOptionsUiToPresentation } from "./Utils";
import IPresentationTreeDataProvider from "./IPresentationTreeDataProvider";

/**
 * Presentation Rules-driven tree data provider.
 */
export default class PresentationTreeDataProvider implements IPresentationTreeDataProvider {
  private _rulesetId: string;
  private _connection: IModelConnection;

  /**
   * Constructor.
   * @param connection Connection to an imodel to pull data from.
   * @param rulesetId Id of the ruleset used by this data provider.
   */
  public constructor(connection: IModelConnection, rulesetId: string) {
    this._rulesetId = rulesetId;
    this._connection = connection;
  }

  /** Id of the ruleset used by this data provider */
  public get rulesetId(): string { return this._rulesetId; }

  /** [[IModelConnection]] used by this data provider */
  public get connection(): IModelConnection { return this._connection; }

  /** Called to get extended options for node requests */
  private createRequestOptions(): HierarchyRequestOptions<IModelConnection> {
    return {
      imodel: this._connection,
      rulesetId: this._rulesetId,
    };
  }

  /**
   * Returns a [[NodeKey]] from given [[TreeNodeItem]].
   * **Warning:** the `node` must be created by this data provider.
   */
  public getNodeKey(node: TreeNodeItem): NodeKey {
    return node.extendedData.key as NodeKey;
  }

  /**
   * Returns nodes
   * @param parentNode The parent node to return children for.
   * @param pageOptions Information about the requested page of data.
   */
  public getNodes = _.memoize(async (parentNode?: TreeNodeItem, pageOptions?: PageOptions): Promise<DelayLoadedTreeNodeItem[]> => {
    if (parentNode) {
      const parentKey = this.getNodeKey(parentNode);
      const childNodes = await Presentation.presentation.getChildren({ ...this.createRequestOptions(), paging: pageOptionsUiToPresentation(pageOptions) }, parentKey);
      return createTreeNodeItems(childNodes, parentNode.id);
    }
    const rootNodes = await Presentation.presentation.getRootNodes({ ...this.createRequestOptions(), paging: pageOptionsUiToPresentation(pageOptions) });
    return createTreeNodeItems(rootNodes);
  }, MemoizationHelpers.getNodesKeyResolver);

  /**
   * Returns the total number of nodes
   * @param parentNode The parent node to return children count for.
   */
  public getNodesCount = _.memoize(async (parentNode?: TreeNodeItem): Promise<number> => {
    if (parentNode) {
      const parentKey = this.getNodeKey(parentNode);
      return await Presentation.presentation.getChildrenCount(this.createRequestOptions(), parentKey);
    } else {
      return await Presentation.presentation.getRootNodesCount(this.createRequestOptions());
    }
  }, MemoizationHelpers.getNodesCountKeyResolver);

  /**
   * Returns filtered node paths.
   * @param filter Filter.
   */
  public getFilteredNodePaths = async (filter: string): Promise<NodePathElement[]> => {
    return Presentation.presentation.getFilteredNodePaths(this.createRequestOptions(), filter);
  }
}

class MemoizationHelpers {
  public static createKeyForPageOptions(pageOptions?: PageOptions) {
    if (!pageOptions)
      return "0/0";
    return `${(pageOptions.start) ? pageOptions.start : 0}/${(pageOptions.size) ? pageOptions.size : 0}`;
  }
  public static createKeyForTreeNodeItem(item?: TreeNodeItem) { return item ? item.id : ""; }
  public static getNodesKeyResolver(parent?: TreeNodeItem, pageOptions?: PageOptions) {
    return `${MemoizationHelpers.createKeyForTreeNodeItem(parent)}/${MemoizationHelpers.createKeyForPageOptions(pageOptions)}`;
  }
  public static getNodesCountKeyResolver(parent?: TreeNodeItem) { return MemoizationHelpers.createKeyForTreeNodeItem(parent); }
}
