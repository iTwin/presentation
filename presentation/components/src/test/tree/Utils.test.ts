/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
/* tslint:disable:no-direct-imports */

import { expect } from "chai";
import * as faker from "faker";
import { createRandomECInstanceNode } from "@bentley/presentation-common/lib/test/_helpers/random";
import { PageOptions } from "@bentley/ui-components";
import { createTreeNodeItem, createTreeNodeItems, pageOptionsUiToPresentation } from "../../tree/Utils";

describe("Utils", () => {

  describe("createTreeNodeItem", () => {

    it("creates tree node", () => {
      const node = createRandomECInstanceNode();
      const treeNode = createTreeNodeItem(node);
      expect(treeNode).to.matchSnapshot();
    });

    it("creates tree node with extended data", () => {
      const node = { ...createRandomECInstanceNode(), extendedData: { test: "value" } };
      const treeNode = createTreeNodeItem(node);
      expect(treeNode.extendedData!.test).to.eq("value");
    });

    it("creates tree node with parent id", () => {
      const node = createRandomECInstanceNode();
      const parentId = faker.random.word();
      const treeNode = createTreeNodeItem(node, parentId);
      expect(treeNode).to.matchSnapshot();
    });

    it("creates tree node with custom label styles", () => {
      const node = createRandomECInstanceNode();
      node.fontStyle = "Bold Italic";
      const treeNode = createTreeNodeItem(node);
      expect(treeNode).to.matchSnapshot();
    });

    it("creates tree node without labelDefinition", () => {
      const node = createRandomECInstanceNode();
      node.labelDefinition = undefined;
      const treeNode = createTreeNodeItem(node);
      expect(treeNode).to.matchSnapshot();
    });

  });

  describe("createTreeNodeItems", () => {
    it("creates tree nodes", () => {
      const nodes = [createRandomECInstanceNode(), createRandomECInstanceNode()];
      const treeNode = createTreeNodeItems(nodes);
      expect(treeNode).to.matchSnapshot();
    });

    it("creates tree nodes with parentId", () => {
      const nodes = [createRandomECInstanceNode(), createRandomECInstanceNode()];
      const parentId = faker.random.word();
      const treeNode = createTreeNodeItems(nodes, parentId);
      expect(treeNode).to.matchSnapshot();
    });

  });

  describe("pageOptionsUiToPresentation", () => {
    it("returns undefined if passed undefined parameter", () => {
      const result = pageOptionsUiToPresentation(undefined);
      expect(result).to.be.equal(undefined);
    });

    it("converts ui page options to presentation page options", () => {
      const size = faker.random.number();
      const start = faker.random.number();
      const pageOptions: PageOptions = { size, start };
      const result = pageOptionsUiToPresentation(pageOptions);

      expect(result).to.not.be.undefined;
      expect(result!.size).to.be.equal(size);
      expect(result!.start).to.be.equal(start);
    });

  });

});
