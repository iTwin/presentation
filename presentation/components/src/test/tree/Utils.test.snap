// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Utils createTreeNodeItem creates tree node 1`] = `
Object {
  "__presentation-components/key": Object {
    "instanceKey": Object {
      "className": "Fantastic Rubber Towels",
      "id": "0x44780000014325",
    },
    "pathFromRoot": Array [
      "2a046127-f623-4128-9848-8bc1c9d119fb",
      "96ab449d-f282-409d-8ff6-91955c55e9e3",
    ],
    "type": "ECInstanceNode",
  },
  "autoExpand": true,
  "icon": "AGP",
  "id": "96ab449d-f282-409d-8ff6-91955c55e9e3/2a046127-f623-4128-9848-8bc1c9d119fb",
  "isCheckboxDisabled": true,
  "isCheckboxVisible": true,
  "label": PropertyRecord {
    "property": Object {
      "displayLabel": "Label",
      "name": "node_label",
      "typename": "string",
    },
    "value": Object {
      "displayValue": "Indian Rupee Ngultrum",
      "value": "back up",
      "valueFormat": 0,
    },
  },
  "style": Object {
    "colorOverrides": Object {
      "backgroundColor": -920212993,
      "color": 1737639167,
    },
  },
}
`;

exports[`Utils createTreeNodeItem creates tree node with custom label styles 1`] = `
Object {
  "__presentation-components/key": Object {
    "instanceKey": Object {
      "className": "Missouri",
      "id": "0x1ab100000013b4",
    },
    "pathFromRoot": Array [
      "dc48ca2b-1044-4800-bce7-a4f6d779f7dd",
      "0ef0a772-c3d3-403a-80a8-4a6a8c62fda6",
    ],
    "type": "ECInstanceNode",
  },
  "checkBoxState": 1,
  "hasChildren": true,
  "icon": "ADP",
  "id": "0ef0a772-c3d3-403a-80a8-4a6a8c62fda6/dc48ca2b-1044-4800-bce7-a4f6d779f7dd",
  "isCheckboxDisabled": true,
  "isCheckboxVisible": true,
  "label": PropertyRecord {
    "property": Object {
      "displayLabel": "Label",
      "name": "node_label",
      "typename": "string",
    },
    "value": Object {
      "displayValue": "Mews",
      "value": "regional",
      "valueFormat": 0,
    },
  },
  "style": Object {
    "colorOverrides": Object {
      "backgroundColor": 461301503,
      "color": 1658093567,
    },
    "isBold": true,
    "isItalic": true,
  },
}
`;

exports[`Utils createTreeNodeItem creates tree node with parent id 1`] = `
Object {
  "__presentation-components/key": Object {
    "instanceKey": Object {
      "className": "Shoes",
      "id": "0x1091d0000016b29",
    },
    "pathFromRoot": Array [
      "559da389-bc4e-4ef0-a5d0-cb8b62e60b80",
      "175c270d-a1a3-4ba3-94ef-8866c2a100d7",
    ],
    "type": "ECInstanceNode",
  },
  "description": "Omnis quam est minima voluptas et velit aperiam.",
  "hasChildren": true,
  "icon": "Automotive",
  "id": "175c270d-a1a3-4ba3-94ef-8866c2a100d7/559da389-bc4e-4ef0-a5d0-cb8b62e60b80",
  "label": PropertyRecord {
    "property": Object {
      "displayLabel": "Label",
      "name": "node_label",
      "typename": "string",
    },
    "value": Object {
      "displayValue": "JBOD",
      "value": "South Dakota",
      "valueFormat": 0,
    },
  },
  "parentId": "deposit",
}
`;

exports[`Utils createTreeNodeItem creates tree node without labelDefinition 1`] = `
Object {
  "__presentation-components/key": Object {
    "instanceKey": Object {
      "className": "value-added",
      "id": "0xdb590000003706",
    },
    "pathFromRoot": Array [
      "c7ea16a7-b57b-4973-b341-ed11dd8506c2",
      "9d2bd13d-baba-48d5-a16f-11a6bebd4508",
    ],
    "type": "ECInstanceNode",
  },
  "description": "Commodi quae explicabo hic aut quod rem odit odio saepe.",
  "icon": "Direct",
  "id": "9d2bd13d-baba-48d5-a16f-11a6bebd4508/c7ea16a7-b57b-4973-b341-ed11dd8506c2",
  "label": "calculate",
}
`;

exports[`Utils createTreeNodeItems creates tree nodes 1`] = `
Array [
  Object {
    "__presentation-components/key": Object {
      "instanceKey": Object {
        "className": "e-tailers",
        "id": "0x17b9c000000d4fc",
      },
      "pathFromRoot": Array [
        "61ac32e0-9daf-4f7d-a66e-d5ccf6b5e845",
        "ce78d049-0790-4e52-8185-2841f2ddecb9",
      ],
      "type": "ECInstanceNode",
    },
    "autoExpand": true,
    "description": "Beatae necessitatibus eum ab harum.",
    "hasChildren": true,
    "icon": "Automotive",
    "id": "ce78d049-0790-4e52-8185-2841f2ddecb9/61ac32e0-9daf-4f7d-a66e-d5ccf6b5e845",
    "isCheckboxDisabled": true,
    "isCheckboxVisible": true,
    "label": PropertyRecord {
      "property": Object {
        "displayLabel": "Label",
        "name": "node_label",
        "typename": "string",
      },
      "value": Object {
        "displayValue": "systematic",
        "value": "wireless",
        "valueFormat": 0,
      },
    },
    "style": Object {
      "colorOverrides": Object {
        "backgroundColor": -2024888833,
      },
    },
  },
  Object {
    "__presentation-components/key": Object {
      "instanceKey": Object {
        "className": "invoice",
        "id": "0x159ee000000b364",
      },
      "pathFromRoot": Array [
        "62600815-3750-4f3d-aeab-b1421892b75e",
        "f471908f-cd9c-4af2-9d63-3b21a81337a8",
      ],
      "type": "ECInstanceNode",
    },
    "hasChildren": true,
    "id": "f471908f-cd9c-4af2-9d63-3b21a81337a8/62600815-3750-4f3d-aeab-b1421892b75e",
    "label": PropertyRecord {
      "property": Object {
        "displayLabel": "Label",
        "name": "node_label",
        "typename": "string",
      },
      "value": Object {
        "displayValue": "Assurance",
        "value": "Hong Kong Dollar",
        "valueFormat": 0,
      },
    },
    "style": Object {
      "colorOverrides": Object {
        "backgroundColor": -2105862145,
        "color": 1250650111,
      },
    },
  },
]
`;

exports[`Utils createTreeNodeItems creates tree nodes with parentId 1`] = `
Array [
  Object {
    "__presentation-components/key": Object {
      "instanceKey": Object {
        "className": "Granite",
        "id": "0x9ece000000928e",
      },
      "pathFromRoot": Array [
        "a72a9ca8-debb-43f2-90e0-38b00330fde0",
        "7d2ec992-c8d5-4f77-a6e2-d57cff276d24",
      ],
      "type": "ECInstanceNode",
    },
    "checkBoxState": 1,
    "id": "7d2ec992-c8d5-4f77-a6e2-d57cff276d24/a72a9ca8-debb-43f2-90e0-38b00330fde0",
    "isCheckboxVisible": true,
    "label": PropertyRecord {
      "property": Object {
        "displayLabel": "Label",
        "name": "node_label",
        "typename": "string",
      },
      "value": Object {
        "displayValue": "Harbors",
        "value": "eyeballs",
        "valueFormat": 0,
      },
    },
    "parentId": "one-to-one",
    "style": Object {
      "colorOverrides": Object {
        "backgroundColor": 479940351,
        "color": 615703295,
      },
    },
  },
  Object {
    "__presentation-components/key": Object {
      "instanceKey": Object {
        "className": "exuding",
        "id": "0x17679000001797b",
      },
      "pathFromRoot": Array [
        "29db17a7-2702-4f3a-b980-5abe6ed9fcd7",
        "bda1c500-d4d5-4d63-809e-4f2bdad7ba2e",
      ],
      "type": "ECInstanceNode",
    },
    "autoExpand": true,
    "checkBoxState": 1,
    "hasChildren": true,
    "icon": "methodologies",
    "id": "bda1c500-d4d5-4d63-809e-4f2bdad7ba2e/29db17a7-2702-4f3a-b980-5abe6ed9fcd7",
    "isCheckboxDisabled": true,
    "isCheckboxVisible": true,
    "label": PropertyRecord {
      "property": Object {
        "displayLabel": "Label",
        "name": "node_label",
        "typename": "string",
      },
      "value": Object {
        "displayValue": "Cambridgeshire",
        "value": "Sharable",
        "valueFormat": 0,
      },
    },
    "parentId": "one-to-one",
    "style": Object {
      "colorOverrides": Object {
        "backgroundColor": 632996863,
      },
    },
  },
]
`;
